### PR TITLE
brianyin/ajs-97-refactor-chatcontext-to-lk-agent-v1

### PR DESCRIPTION
**Pull request by toubatbrian**

### CODE REVIEW

File: agents/src/llm/misc.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import { v4 as uuidv4 } from 'uuid';
+
+export function shortuuid(prefix: string): string {
+  return `${prefix}_${uuidv4().slice(0, 12)}`;
+}
```
Comment (Shubhrakanti) [inline agents/src/llm/misc.ts]: ~~Let's do this in a separate PR. I have a ticket for this. (AJS-60)~~

File: agents/src/llm/provider_format/openai.test.ts [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+import { serializeImage } from '../utils.js';
+import { toChatCtx } from './openai.js';
+
+// Mock the serializeImage function
+vi.mock('../utils.js', () => ({
+  serializeImage: vi.fn(),
+}));
+
+describe('toChatCtx', () => {
+  const serializeImageMock = vi.mocked(serializeImage);
```
Comment (Shubhrakanti) [inline agents/src/llm/provider_format/openai.test.ts:17]: Personally I'm a bigger fan of having fewer tests that test really the edge cases and explain non intuitive behavior - what do you think?

**Review by Shubhrakanti — APPROVED** [review]
This PR is over 2000 lines of code making it pretty hard to review.


I just took a brief glance and seems fine.

**Comment by Shubhrakanti** [comment]
>  The changes appear quite large, but the actual impact is smaller than it seems.
> 
> Most of the updates are related to refactoring, such as method call adjustments, which is either 0% or 100%. A significant portion of the diff consists of unit tests. The actual new code is minimal and primarily introduces a new approach for converting the LiveKit 1.0 chat context to the OpenAI format.
> 

My recommendation is then you make one PR with all the functional changes and a follow up PR that actually implements those changes/add tests. Or at least add a comment highlighting the functional changes. Otherwise the reviewer has to drudge through all the code to figure out what the core changes are.

File: agents/src/llm/misc.ts (hunk 1 above)
Comment (Shubhrakanti) [reply agents/src/llm/misc.ts]: I think it's fine for now, but can you add a comment on this function saying `//TODO(AJS-60) refactor all calls to randomUUID to use this`

File: agents/src/llm/utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+      // Decode and verify the pattern
+      const { decodedImage } = await decodeImageToRaw(result.base64Data!);
+      const decoded = decodedImage.data;
+
+      // Check each pixel - PNG is lossless so values should be exact
+      expectPixel(decoded, 0, { r: 255, g: 0, b: 0, a: 255 }); // Top-left (red)
+      expectPixel(decoded, 4, { r: 0, g: 255, b: 0, a: 255 }); // Top-right (green)
+      expectPixel(decoded, width * 4, { r: 0, g: 0, b: 255, a: 255 }); // Bottom-left (blue)
+      expectPixel(decoded, (width + 1) * 4, { r: 255, g: 255, b: 255, a: 255 }); // Bottom-right (white)
+    });
```
Comment (Shubhrakanti) [inline agents/src/llm/utils.test.ts]: It's fine to have AI generate the tests, but it tends to have overly verbose comments. I think as you go through the tests that it writes you can go thoughts and remove them.

File: agents/src/voice/audio_recognition.ts [hunk 4]
```diff
@@ -190,7 +190,7 @@ export class AudioRecognition {
     }
 
     chatCtx = chatCtx.copy();
-    chatCtx.append({ role: ChatRole.USER, text: this.audioTranscript });
+    chatCtx.addMessage({ role: 'user', content: this.audioTranscript });
```
Comment (Shubhrakanti) [inline agents/src/voice/audio_recognition.ts:193]: I think the enum is better? I guess the python side uses raw strings so we're trying to keep it the same?

I would at least add a follow up ticket to think about making this an enum.

File: package.json [hunk 5]
```diff
@@ -40,11 +41,11 @@
     "eslint-plugin-standard": "^5.0.0",
     "eslint-plugin-tsdoc": "^0.2.17",
     "prettier": "^3.2.5",
+    "tsup": "^8.4.0",
     "turbo": "^1.13.3",
     "typedoc": "^0.25.13",
     "typescript": "^5.4.5",
-    "vitest": "^3.2.2",
-    "tsup": "^8.4.0"
+    "vitest": "^3.2.2"
```
Comment (Shubhrakanti) [inline package.json:48]: no-op same with `"@types/node": "^22.13.10",`. Did prettier move these lines?

Either way, it's good to go through the PR and minimize no-op line changes if possible or comment why they happened.

File: agents/src/llm/provider_format/openai.test.ts (hunk 2 above)
Comment (toubatbrian) [reply agents/src/llm/provider_format/openai.test.ts:17]: Agree. These tests are generally test the core path & normal cases. I'm putting this test file because this function is part of the critical path of the openai's `toChatCtx` function. 

We should definitely make test sets smaller and add edge case testing only when necessary in the future

File: package.json (hunk 5 above)
Comment (toubatbrian) [reply package.json:48]: I guess when I install a new package by pnpm install, it automatically re-format the order of dependency. Will make sure I make a comment on these details in the following PRs.

File: agents/src/llm/provider_format/openai.test.ts (hunk 2 above)
Comment (Shubhrakanti) [reply agents/src/llm/provider_format/openai.test.ts:17]: Yeah. Honestly I should be writing more tests too. helps explain behavior.

File: package.json (hunk 5 above)
Comment (Shubhrakanti) [reply package.json:48]: Sounds good. Main thing I look for is making sure there are no version changes in existing packages that I should be aware of.

File: agents/src/llm/utils.test.ts (hunk 3 above)
Comment (toubatbrian) [reply agents/src/llm/utils.test.ts]: Sounds good! btw, I’ve also reviewed all the AI-generated tests, they should cover the intended behaviors.

File: agents/src/voice/audio_recognition.ts (hunk 4 above)
Comment (toubatbrian) [reply agents/src/voice/audio_recognition.ts:193]: The strings here are type-safe too, in case that’s a main concern. I feel like using strings makes more sense, since it allows users to construct the message payload just like they would when passing arguments to OpenAI’s chat completion calls—no need to import anything extra.

Comment (Shubhrakanti) [reply agents/src/voice/audio_recognition.ts:193]: Ohhh I see. Yeah let's def keep it a string here.

File: agents/src/llm/provider_format/index.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+// SPDX-FileCopyrightText: 2025 LiveKit, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+import type { ChatContext } from '../chat_context.js';
+import { toChatCtx as toChatCtxOpenai } from './openai.js';
+
+export type ProviderFormat = 'openai';
+
+export async function toChatCtx(
+  format: ProviderFormat,
```
Comment (Shubhrakanti) [inline agents/src/llm/provider_format/index.ts:10]: @toubatbrian - this can just go in `/llm/utils.ts` no? misc and utils might be confusing.
